examples/login-with-role/src/main.rs

```diff
@@ -0,0 +1,128 @@
+//! ```
+
+use std::{collections::HashMap, sync::Arc};
+
+use async_trait::async_trait;
+use axum::{
+    extract::{FromRequest, RequestParts},
+    http::StatusCode,
+    response::IntoResponse,
+    routing::get,
+    Extension, Router,
+};
+use axum_login::{
+    axum_sessions::{async_session::MemoryStore as SessionMemoryStore, SessionLayer},
+    memory_store::MemoryStore as AuthMemoryStore,
+    AuthLayer, AuthUser, RequireAuthorizationLayer,
+};
+use rand::Rng;
+use tokio::sync::RwLock;
+
+#[derive(Debug, Clone, PartialEq)]
+enum Role {
+    User,
+    Admin,
+}
+
+#[derive(Debug, Clone)]
+struct User {
+    id: usize,
+    password_hash: String,
+    role: Role,
+    name: String,
+}
+
+impl User {
+    fn get_rusty_user() -> Self {
+        Self {
+            id: 1,
+            name: "Ferris the Crab".to_string(),
+            password_hash: "password".to_string(),
+            role: Role::Admin,
+        }
+    }
+}
+
+impl AuthUser<Role> for User {
+    fn get_id(&self) -> String {
+        format!("{}", self.id)
+    }
+
+    fn get_password_hash(&self) -> String {
+        self.password_hash.clone()
+    }
+
+    fn get_role(&self) -> Option<Role> {
+        Some(self.role.clone())
+    }
+}
+
+type AuthContext = axum_login::extractors::AuthContext<User, AuthMemoryStore<User>, Role>;
+
+#[tokio::main]
+async fn main() {
+    let secret = rand::thread_rng().gen::<[u8; 64]>();
+
+    let session_store = SessionMemoryStore::new();
+    let session_layer = SessionLayer::new(session_store, &secret).with_secure(false);
+
+    let store = Arc::new(RwLock::new(HashMap::default()));
+    let user = User::get_rusty_user();
+
+    store.write().await.insert(user.get_id(), user);
+
+    let user_store = AuthMemoryStore::new(&store);
+    let auth_layer = AuthLayer::new(user_store, &secret);
+
+    async fn login_handler(mut auth: AuthContext) {
+        auth.login(&User::get_rusty_user()).await.unwrap();
+    }
+
+    async fn logout_handler(mut auth: AuthContext) {
+        dbg!("Logging out user: {}", &auth.current_user);
+        auth.logout().await;
+    }
+
+    async fn protected_handler(Extension(user): Extension<User>) -> impl IntoResponse {
+        format!("Logged in as: {}", user.name)
+    }
+
+    async fn admin_handler(Extension(user): Extension<User>) -> impl IntoResponse {
+        format!("Admin logged in as: {}", user.name)
+    }
+
+    async fn user_handler(Extension(user): Extension<User>) -> impl IntoResponse {
+        format!("User logged in as: {}", user.name)
+    }
+
+    let app = Router::new()
+        .route(
+            "/protected",
+            get(protected_handler).layer(RequireAuthorizationLayer::<User, Role>::login()),
+        )
+        .route(
+            "/protected_admin",
+            get(admin_handler).layer(RequireAuthorizationLayer::<User, Role>::login_with_role(
+                Role::Admin,
+            )),
+        )
+        .route(
+            "/protected_user",
```

Comment:
One thing to note here is that roles aren't ranked or otherwise part of a range which means that the `Role::Admin` variant won't be accepted (only `Role::User`).

If we did want enable authorization sets, where more than one role type might be authorized for a single endpoint, we might consider if [`RangeBounds`](https://doc.rust-lang.org/std/ops/trait.RangeBounds.html) could be helpful.

Reply:
@maxcountryman nice observation and a very good idea. I'll create a task to require an `impl` for the `role` param.

Reply:
Here's one possibility: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0825788e9d4f510700f6d92653988ebd

I think this would just require `PartialOrd`.